### PR TITLE
[MRG] Added read/write methods and tests for dipole

### PR DESCRIPTION
**Pull request by raj1701**
- Added write method for dipole
- Added read method for dipole
- Added required tests

### CODE REVIEW

File: hnn_core/dipole.py [hunk 1]
```diff
@@ -274,13 +333,16 @@ class Dipole(object):
...
         self.times = np.array(times)
 
-        if data.ndim == 1:
-            data = data[:, None]
-
-        if data.shape[1] == 3:
-            self.data = {'agg': data[:, 0], 'L2': data[:, 1], 'L5': data[:, 2]}
-        elif data.shape[1] == 1:
-            self.data = {'agg': data[:, 0]}
+        if type(data) is dict:
```
Comment (jasmainak) [inline hnn_core/dipole.py:336]: use `isinstance`

File: hnn_core/tests/test_dipole.py [hunk 2]
```diff
@@ -49,14 +50,40 @@ def test_dipole(tmpdir, run_hnn_core_fixture):
...
         assert_allclose(dipole_read.data[dpl_key],
                         dipole.data[dpl_key], rtol=0, atol=0.000051)
 
+    # Test IO for hdf5 files
+    dipole.write(dpl_out_hdf5_fname)
+    dipole_read_hdf5 = read_dipole(dpl_out_hdf5_fname)
+    assert_allclose(dipole_read_hdf5.times, dipole.times, rtol=0, atol=0.00051)
+    for dpl_key in dipole.data.keys():
+        assert_allclose(dipole_read_hdf5.data[dpl_key],
+                        dipole.data[dpl_key], rtol=0, atol=0.000051)
```
Comment (jasmainak) [inline hnn_core/tests/test_dipole.py:70]: we should move away from using `atol` and use `rtol` ... it's more robust I believe

File: hnn_core/tests/test_dipole.py [hunk 3]
```diff
@@ -49,14 +50,40 @@ def test_dipole(tmpdir, run_hnn_core_fixture):
...
+
+    # Testing for wrong extension provided
+    with pytest.raises(NameError,
+                       match="File extension should be either txt or hdf5"):
+        dipole.write(tmpdir.join('dpl.xls'))
+
+    # Testing File Not Found Error
+    with pytest.raises(FileNotFoundError,
+                       match="File not found at "):
+        read_dipole(tmpdir.join('dpl1.hdf5'))
```
Comment (jasmainak) [inline hnn_core/tests/test_dipole.py:85]: can you also add a test to check that `Path` object works

**Comment by jasmainak** [comment]
Looks pretty clean otherwise!

**Comment by raj1701** [comment]
Sure will make these changes.

File: hnn_core/dipole.py [hunk 4]
```diff
@@ -611,3 +677,59 @@ def write(self, fname):
...
+        Parameters
+        ----------
+        fname : str | Path object
+            Full path to the output file (.txt or .hdf5)
+
+        Outputs
+        -------
+        A tab separatd txt file where rows correspond
+            to samples and columns correspond to
+            1) time (s),
```
Comment (ntolley) [inline hnn_core/dipole.py:713]:
```suggestion
            1) time (ms),
```

**Review by ntolley — APPROVED** [review]
Great job @raj1701! Happy to merge once you address the lingering comments

File: hnn_core/tests/test_dipole.py (hunk 2 above)
Comment (ntolley) [reply hnn_core/tests/test_dipole.py]: https://stackoverflow.com/questions/65909842/what-is-rtol-for-in-numpys-allclose-function

Checks out! The dipole varies pretty wildly in magnitude between subthreshold and suprathreshold simulations, so considering the precision of `floats` gets worse with bigger numbers `rtol` seems particularly appropriate

**Comment by raj1701** [comment]
@jasmainak @ntolley I have addressed all comments. Please review once

File: hnn_core/dipole.py [hunk 5]
```diff
@@ -593,12 +655,16 @@ def write(self, fname):
...
         A tab separatd txt file where rows correspond
             to samples and columns correspond to
-            1) time (s),
+            1) time (ms),
             2) aggregate current dipole (scaled nAm),
             3) L2/3 current dipole (scaled nAm), and
             4) L5 current dipole (scaled nAm)
         """
 
+        warnings.warn('Writing dipole to txt file is deprecated '
```
Comment (jasmainak) [inline hnn_core/dipole.py:664]: is this tested?

**Comment by jasmainak** [comment]
Thinking about this, @raj1701 I think these changes should be documented in the API section of the `whats_new`. Since the hdf5 format is technically the default now

**Comment by jasmainak** [comment]
Looks good otherwise. Maybe @ntolley can merge once the `whats_new` is updated

File: hnn_core/dipole.py [hunk 6]
```diff
@@ -611,3 +677,59 @@ def write(self, fname):
...
+        A hdf5 file containing the dipole object
+        """
+        print(f'Writing file {fname}')
+        dpl_data = dict()
+        dpl_data['times'] = self.times
+        dpl_data['sfreq'] = self.sfreq
+        dpl_data['nave'] = self.nave
+        dpl_data['data'] = self.data
+        dpl_data['scale_applied'] = self.scale_applied
+        write_hdf5(fname, dpl_data, overwrite=True)
```
Comment (rythorpe) [inline hnn_core/dipole.py:699]: Should we be including a flag in the hdf5 file that specifies the type of HNN-core object that this should be used to create when read back in (i.e., to help prevent someone from accidentally trying to read in file that is incompatible with the `read_dipole()` method?

Comment (raj1701) [reply hnn_core/dipole.py:699]: Yes sure, I will add this.

Comment (rythorpe) [reply hnn_core/dipole.py:699]: If you'd like to save this for later, that's fine too since I just noticed that this PR is already set to auto-merge!

**Comment by raj1701** [comment]
Hey @ntolley @jasmainak @rythorpe, I added a check to verify `object_type` of the object stored/read

**Comment by rythorpe** [comment]
> Hey @ntolley @jasmainak @rythorpe, I added a check to verify `object_type` of the object stored/read

Looks good @raj1701!

**Comment by jasmainak** [comment]
I went ahead and merged since the linkcheck is unrelated to the PR ... so that @raj1701 can move ahead with his other PRs. Thx @raj1701 !

**Pull request merged**